File: 238. Product of Array Except Self/238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int pro = 1, n = nums.size(), zpos;
        bool zero = false;
        vector<int> ans(n,0);
        for(int i = 0; i < n; ++i){
            if(nums[i] == 0){
                if(zero)    return ans;
                else{
                    zero = true;
                    zpos = i;
                }
            }
            else    pro *= nums[i];
        }
        if(zero)    ans[zpos] = pro;
        else
            for(int i = 0; i < n; ++i)
                ans[i] = pro/nums[i];
        return ans;
    }
};

File: 238. Product of Array Except Self/238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length, prefix = 1, suffix = 1;
        int[] ans = new int[n];
        Arrays.fill(ans,1);
        for(int i = 0; i < n; ++i){
            ans[i] *= prefix;
            prefix *= nums[i];
            ans[n-1-i] *= suffix;
            suffix *= nums[n-1-i];
        }
        return ans;
    }
}